File: site-online/web/objects/sitepages/buy.ml
<div class="index-container individual">
  <div class="header agora">
    <div class="bg-image"><img src="/img/agora/space_dark_blue_planet2.png"></div>
    <div class="logo"><img src="/img/agora/agoralogo.png"></div>
  </div>

  <div class="nft-content">
    <div class="back-button">
      <button id = "backButton"><div class="capitalise">Back</div></button>
    </div>

    <div class="nft-details">
      <div class="nft-intro">
        <!-- div class="bg-sparkle invisible"><img src="/img/agora/glow_v01.png"></div -->
        <div id="nftimage" class="nft-card-image loading"></div>
        <div class="nft-basic">
          <p class="capitalise label">Art by</p>
          <p id="artistName" class="name loading">&nbsp;</p>
          <p id="artistTitle" class="basic-title loading">&nbsp;</p>
          <p id="artistSource" class="source loading">&nbsp;</p>
        </div>
      </div>

      <div class="nft-detail">
        <div class="title collection-title">
          <h2 id="collectionName" class="capitalise collection loading">&nbsp;</h2>
					<div class="social-list">
						<div class="social-icon twitter">
							<a href="http://go.jedstar.com/twitter" aria-label="twitter">
							<img src="https://assets.jedstar.space/img/Twitter_footer.webp" alt="twitter">
							</a>
						</div>
						<div class="social-icon facebook">
							<a href="http://go.jedstar.com/facebook" aria-label="facebook">
							<img src="https://assets.jedstar.space/img/facebook_footer.webp" alt="facebook">
							</a>
						</div>
						<div class="social-icon reddit">
							<a href="http://go.jedstar.com/reddit" aria-label="reddit">
							<img src="https://assets.jedstar.space/img/reddit_footer.webp" alt="reddit">
							</a>
						</div>
					</div>
        </div>

        <div class="rainbow-border"></div>

        <div class="nft-detail-content">
          <div class="nft-detail-name">
            <p class="capitalise label">Name</p>
            <div class="title">
              <h2 id="nftname" class="capitalise loading paragraph">&nbsp;</h2>
            </div>
            <div class="card-rarity">
              <p id="rarity-label" class="capitalise loading paragraph">&nbsp;</p>
            </div>
          </div>

					<div class="nft-respec">
						<p class="capitalise label">Current Price</p>
						<div class="agora-buy-price">
              <div class="pricecurrency-wrapper"><select id="pricecurrency"><option value="USDC">USDC</option></select></div>
							<h3 id="price" class="loading">&nbsp;</h3>
						</div>
						<p id="mintedcount" class="remaining capitalise loading">&nbsp;</p>
						<button id="buynowsolo" class="buynow"><div class="capitalise">Buy now</div></button>
					</div>

          <div class="nft-detail-description">
            <p class="capitalise label">Description</p>
            <p id="nft-description-text" class="loading">&nbsp;</p>
          </div>

          <div class="nft-detail-abilities">
            <p class="capitalise label">Attributes & utilities</p>
            <p id="nft-utilities-text" class="loading"></p><br>
          </div>

          <div class="nft-detail-attributes">
            <div class="title">
              <h2 class="capitalise">Attributes</h2>
            </div>
            <div class="abilities-respec">
              <p class="capitalise">Upon minting <span id ="attributes-rarity">...</span> NFTS are given <span id ="attributes-given"><span id = "attributes-given-num">0</span> of 5</span> random attributes that have a random power scale from 1-100.</p>
            </div>
            <div class="attributes-table">
              <table>
                <tr id="Surge"><td class="attr-bar"><p class="capitalise">Serge</p></td><td class="attr-val">?</td></tr>
                <tr id="Might"><td class="attr-bar"><p class="capitalise">Might</p></td><td class="attr-val">?</td></tr>
                <tr id="Resilience"><td class="attr-bar"><p class="capitalise">Resilience</p></td><td class="attr-val">?</td></tr>
                <tr id="Wit"><td class="attr-bar"><p class="capitalise">Wit</p></td><td class="attr-val">?</td></tr>
                <tr id="Luck"><td class="attr-bar"><p class="capitalise">Luck</p></td><td class="attr-val">?</td></tr>
              </table>
            </div>
            <div class="chainlink">
              <p>Powered by</p>
              <img src="/img/agora/chainlink.png">
              <p class="capitalise"><a href="https://chain.link/vrf" target="_blank" rel="nofollow">about</a></p>
            </div>
          </div>
          <div><p>This NFT offers utility across the JEDSTAR universe of apps and games. </p></div>
          <div class="nft-detail-silvervolt">
            <div class="nft-detail-logo"><img src="https://assets.jedstar.space/img/SILVERVOLT_LOGO.webp"></div>
            <p>This NFT offers various in app Boosts and Upgrades to our SILVERVOLT gaming product. Sign up and top off your favorite games for FREE at <a href="">www.silvervolt.app</a></p>
            <div class="nft-detail-boosts">
              <div class="boost off">
                <div class="boost-check"><img src="/img/agora/check.svg"></div>
                <div class="boost-title"><p class="capitalise"><span id="boost-bid"></span> bid boost</p></div>
                <div class="boost-subtitle"><p class="label">Gives you additional weekly auction bids. Up to <span id = "numBid">...</span> / week for a <span class = "attributes-rarities">...</span>.</p></div>
                <div class="boost-stat"><p class="capitalise">Serge</p></div>
              </div>

              <div class="boost off">
                <div class="boost-check"><img src="/img/agora/check.svg"></div>
                <div class="boost-title"><p class="capitalise"><span id="boost-mining"></span> mining boost</p></div>
                <div class="boost-subtitle"><p class="label">Boosts your VOLT mining power. Up to <span id = "numMining">...</span>% for a <span class = "attributes-rarities">...</span>.</p></div>
                <div class="boost-stat"><p class="capitalise">Might</p></div>
              </div>

              <div class="boost off">
                <div class="boost-check"><img src="/img/agora/check.svg"></div>
                <div class="boost-title"><p class="capitalise">Tier boost</p></div>
                <div class="boost-subtitle"><p class="label">Instant access to <span id = "tierType">...</span>-Tier Auctions for a <span class = "attributes-rarities">...</span>.</p></div>
                </div>

              <div class="boost off">
                <div class="boost-check"><img src="/img/agora/check.svg"></div>
                <div class="boost-title"><p class="capitalise"><span id="boost-referral"></span> Referral boost</p></div>
                <div class="boost-subtitle"><p class="label">Boosts your referral VOLT bonus. Up to <span id = "numReferral">...</span>% extra for a <span class = "attributes-rarities">...</span>.</p></div>
                <div class="boost-stat"><p class="capitalise">Resilience</p></div>
              </div>

              <div class="boost off">
                <div class="boost-check"><img src="/img/agora/check.svg"></div>
                <div class="boost-title"><p class="capitalise"><span id="boost-wheel"></span> Wheel Spin Boost</p></div>
                <div class="boost-subtitle"><p class="label">Cheaper Wheel Spins! (Coming Soon) Up to <span id = "numWheel">...</span>% off wheel spins for a <span class = "attributes-rarities">...</span>.</p></div>
                <div class="boost-stat"><p class="capitalise">Wit</p></div>
              </div>

              <div class="boost off">
                <div class="boost-check"><img src="/img/agora/check.svg"></div>
                <div class="boost-title"><p class="capitalise"><span id="boost-lottery"></span> Lottery Ticket Boost</p></div>
                <div class="boost-subtitle"><p class="label">Cheaper Lottery Tickets! Up to <span id = "numLottery">...</span>% off Lottery tickets for a <span class = "attributes-rarities">...</span>.</p></div>
                <div class="boost-stat"><p class="capitalise">Luck</p></div>
              </div>

              <div class="boost off" id = "highRoller">
                <div class="boost-check"><img src="/img/agora/check.svg"></div>
                <div class="boost-title"><p class="capitalise">High roller access</p></div>
                <div class="boost-subtitle"><p class="label">Coming Soon.</p></div>
              </div>

              <div class="boost off">
                <div class="boost-check"><img src="/img/agora/check.svg"></div>
                <div class="boost-title"><p class="capitalise">Custom name</p></div>
                <div class="boost-subtitle"><p class="label">Update your handle to be a custom name! Ultra-Rare NFTs and above.</p></div>
              </div>

              <div class="boost off">
                <div class="boost-check"><img src="/img/agora/check.svg"></div>
                <div class="boost-title"><p class="capitalise">Custom avatar</p></div>
                <div class="boost-subtitle"><p class="label">Select your own custom Avatar! Rare NFTs and above.</p></div>
              </div>

            </div>
          </div>
          <div class="nft-detail-extras">
            <div class="nft-detail-extra">
              <div class="nft-detail-logo"><img src="/img/agora/starstaking_logo_transparent_horizontal_v01.png"></div>
              <p>JEDSTAR Gaming’s official staking platform</p>
              <p class="capitalise label">Coming soon</p>
            </div>

            <div class="nft-detail-extra">
              <div class="nft-detail-logo"><img src="/img/agora/nightclaws.png"></div>
              <p>Arcade RPG Game (Mobile)</p>
              <p class="capitalise label">Coming soon</p>
            </div>

            <div class="nft-detail-extra">
              <div class="nft-detail-logo"><img src="/img/agora/ascension.png"></div>
              <p>Real-Time Strategy Game (Mobile)</p>
              <p class="capitalise label">Coming soon</p>
            </div>

            <div class="nft-detail-extra">
              <div class="nft-detail-logo"><img src="/img/agora/aethernova.png"></div>
              <p>Collectible Card Game (Mobile/PC)</p>
              <p class="capitalise label">Coming soon</p>
            </div>
          </div>

        </div>
      </div>
    </div>
  </div>
</div>

File: site-online/web/content/js/buy.js
let x = window.location.toString().split('/');
if (x[x.length-2] == "buy" || x[x.length-2] == "?c=buy"){
  let productId = x[x.length-1] * 1;
  let product_purchase_details = {};
  Agora.on("ready").then(function(){
    const share_buttons = document.querySelectorAll('.share-icon');
    share_buttons.forEach(button => {
      button.addEventListener('click', function(event){
        let icons = event.target.parentNode.id;
        let URL;
        switch(icons){
          case 'twitter':
            URL = "https://twitter.com/intent/tweet?url=" + window.location.href;
            break;
          case 'facebook':
            URL = "https://www.facebook.com/sharer/sharer.php?u=" + window.location.href;
            break;
          case 'reddit':
            URL = "https://reddit.com/submit?url=" + window.location.href;
            break;
          case 0:
            break;
        }
        if (URL) window.open(URL, '_blank').focus();
      });
    })
    const back_button = document.getElementById("backButton");
    back_button.addEventListener('click', function(){
      history.back();
    });
    chainRead.minted_count(productId).then(function(minted_qty){
      chainRead.get_mint_limits(productId).then(function(mint_limits){
        let mintedLbl = document.getElementById("mintedcount");
        if (minted_qty/mint_limits >= 0.75){
          mintedLbl.innerHTML = (mint_limits - minted_qty)+" of "+mint_limits+" remain";
        }else{
          mintedLbl.innerHTML = minted_qty + " minted";
        }
        mintedLbl.classList.remove("loading");
        mintedLbl.classList.add("fade-in");
      });
    }).catch(function(err){
      console.error(err);
    });

    chainRead.get_product_groupId(productId).then(function(groupId){
      if (groupId < 4){
        chainRead.get_group_name(groupId).then(function(group_name){
          let collectionLbl = document.getElementById("collectionName");
          collectionLbl.innerHTML = "<span class='color-blue'>"+group_name+"</span> collection";
          collectionLbl.classList.remove("loading");
          collectionLbl.classList.add("fade-in");
        }).catch(function(err){
          console.error("Group name retrieval failure", err);
        })
      }else{
        window.location = "/";
      }
    }).catch(function(err){
      console.error("Group ID retrieval failure", err);
    });

    chainRead.get_product_tier(productId).then(function(tierId){
      let rarityLbl = document.getElementById("rarity-label");
      let tierName, tierClass;
      let attributes_rarity = document.getElementById("attributes-rarity");
      let attributes_given_num = document.getElementById("attributes-given-num");
      let bid = document.getElementById("numBid");
      let mining = document.getElementById("numMining");
      let referral = document.getElementById("numReferral");
      let wheel = document.getElementById("numWheel");
      let lottery = document.getElementById("numLottery");
      let tier = document.getElementById("tierType");
      document.getElementById("attributes-given").style.color = "#1C84FE";

      let boosts = document.querySelectorAll(".boost");
      for (let boost of boosts){
        if (boost.id !== "highRoller"){
          boost.classList.remove("off");
        }
      }

      tierId = tierId * 1;
      switch(tierId){
        case 0:
          tierName = "Secret rare";
          tierClass = "secretrare";
          attributes_rarity.style.color = "#F23535";
          attributes_given_num.innerHTML = 5;
          bid.innerHTML = 5, mining.innerHTML = 10, referral.innerHTML = 4, wheel.innerHTML = 80, lottery.innerHTML = 80, tier.innerHTML = "Diamond";
          document.getElementById("highRoller").classList.remove("off");
          break;
        case 1:
          tierName = "Ultra rare";
          tierClass = "ultrarare";
          attributes_rarity.style.color = "#31FF5D";
          attributes_given_num.innerHTML = 4;
          bid.innerHTML = 4, mining.innerHTML = 8, referral.innerHTML = 4, wheel.innerHTML = 70, lottery.innerHTML = 70, tier.innerHTML = "Platinum";
          break;
        case 2:
          tierName = "Rare";
          tierClass = "rare";
          attributes_rarity.style.color = "#FFD131";
          attributes_given_num.innerHTML = 3;
          bid.innerHTML = 3, mining.innerHTML = 6, referral.innerHTML = 2, wheel.innerHTML = 60, lottery.innerHTML = 60, tier.innerHTML = "Gold";
          break;
        case 3:
          tierName = "Uncommon";
          tierClass = "uncommon";
          attributes_rarity.style.color = "#318DFF";
          attributes_given_num.innerHTML = 2;
          bid.innerHTML = 2, mining.innerHTML = 5, referral.innerHTML = 1.5, wheel.innerHTML = 50, lottery.innerHTML = 50, tier.innerHTML = "Silver";
          break;
        case 4:
          tierName = "Common";
          tierClass = "common";
          attributes_rarity.style.color = "#D9D9D9";
          attributes_given_num.innerHTML = 1;
          bid.innerHTML = 1, mining.innerHTML = 4, referral.innerHTML = 1, wheel.innerHTML = 40, lottery.innerHTML = 40, tier.innerHTML = "Not available for Common NFTs";
          break;
        default: //case 5:
          tierName = "Standard";
          tierClass = "common";
          attributes_rarity.style.color = "#D9D9D9";
          attributes_given_num.innerHTML = 1;
          bid.innerHTML = "N/A", mining.innerHTML = "N/A", referral.innerHTML = "N/A", wheel.innerHTML = "N/A", lottery.innerHTML = "N/A", tier.innerHTML = "Not available for Standard NFTs";
          break;
      }
      let attribute_rarities = document.getElementsByClassName("attributes-rarities");
      for (let attribute of attribute_rarities){
        attribute.innerHTML = tierName;
      }
      rarityLbl.innerHTML = tierName;
      attributes_rarity.innerHTML = tierName;
      rarityLbl.classList.add(tierClass);
      rarityLbl.classList.remove("loading");
      rarityLbl.classList.add("fade-in");
    }).catch(function(err){
      console.error("Tier ID retrieval failed", err);
    })

    servercomms.get_product_pricing(productId).then(function(res){
      if (res.status == "ok"){
        let priceLbl = document.getElementById("price");
        let nameLbl = document.getElementById("nftname");
        let imgNft = document.getElementById("nftimage");
        let imgAsset = new Image();
        let btn_buynow = document.getElementById("buynowsolo");
        let descriptionTxt = document.getElementById("nft-description-text");
        let utilTxt = document.getElementById("nft-utilities-text");
        let artistName = document.getElementById("artistName");
        let artistTitle = document.getElementById("artistTitle");
        let artistSource = document.getElementById("artistSource");

        let animFadeIn = function (domElemArr){
          for (let e=0; e < domElemArr.length; e++){
            domElemArr[e].classList.remove("loading");
            domElemArr[e].classList.add("fade-in");
          }
        }

        price.innerHTML = res.price["USDC"];

        nftname.innerHTML = res.name;

        imgAsset.onload = function(){
          imgNft.innerHTML = "<img src='https://assets.jedstar.space/nft/"+productId+".jpg' />";
        };
        imgAsset.src = "https://assets.jedstar.space/nft/"+productId+".jpg";

        let blurbs = JSON.parse(res.description);
        descriptionTxt.innerHTML = blurbs.description;
        // utilTxt.innerHTML = blurbs.utilities;
        utilTxt.innerHTML = "<span><span style = \"font-weight: bold;\">What is an ATTRIBUTE?:</span> When your GENESIS NFT is minted, your Character will be assigned random attribute bonuses, across 5 possible attributes; SERGE, MIGHT, RESILIENCE, WIT and LUCK. Each Attribute will have a bonus value between 1 to 100.</span><br><br>\
              <span><span style = \"font-weight: bold;\">How many ATTRIBUTE Bonuses?</span> A Rarer GENESIS NFT will get more randomly generated Attribute bonuses than a common GENESIS NFT (1 for a Common, 2 for an Uncommon, 3 for a Rare, 4 for an Ultra-Rare and all 5 for a Secret-Rare!). This means that a Common NFT will pick one Attribute at random to allocate your score to, whereas a Secret-Rare will have scores in all 5 Attributes.</span><br><br>\
              <span><span style = \"font-weight: bold;\">Utility:</span> The Attribute scores will directly correlate with increased power, buffs and features in JEDSTAR Games! Scroll down to see what these effects might be.</span>"

        let artBlurb = JSON.parse(res.artist.blurb);
        artistName.innerHTML = res.artist.name;
        artistTitle.innerHTML = artBlurb.title;
        artistSource.innerHTML = artBlurb.source;

        btn_buynow.dataset["productId"] = productId;
        product_purchase_details.pricing = res.price;
        product_purchase_details.name = res.name;

        //Fade all the text blocks in
        animFadeIn([price, nftname, imgNft, descriptionTxt, utilTxt, artistName, artistTitle, artistSource]);

        let currDropdown = document.getElementById("pricecurrency");
        let load_all_prices = function(){
          currDropdown.innerHTML = "";
          for (curr in product_purchase_details.pricing){
            if (product_purchase_details.pricing.hasOwnProperty(curr)){
              if ((curr == "AGORA" && Agora.balances["AGORA"] > 0) || (curr != "AGORA")){
                let opta = document.createElement("OPTION");
                opta.value = curr;
                opta.innerHTML = curr;
                currDropdown.appendChild(opta);
              }
            }
          }
        };
        load_all_prices();
        currDropdown.addEventListener("change", function(){
          priceLbl.innerHTML = product_purchase_details.pricing[currDropdown.value];
          if (typeof product_purchase_details.pricing[currDropdown.value] != "undefined" && Agora.balances[currDropdown.value] >= product_purchase_details.pricing[currDropdown.value]){
            btn_buynow.classList.add('enabled');
          }else{
            btn_buynow.classList.remove('enabled');
          }
        });
        Agora.hook("balances", function(balances){
          //this fires everytime the user's balance is updated
          if (Agora.balances["AGORA"] > 0 ){
            load_all_prices();
          }
          for (curr in Agora.balances){
            if (Agora.balances.hasOwnProperty(curr)){
              if (typeof product_purchase_details.pricing[curr] != "undefined" && Agora.balances[curr] >= product_purchase_details.pricing[curr]){
                for(var i=0; i<currDropdown.options.length; i++) {
                  if ( currDropdown.options[i].value == curr ) {
                    currDropdown.selectedIndex = i;
                    break;
                  }
                }
                btn_buynow.classList.add('enabled');
                break;
              }
            }
          }
          priceLbl.innerHTML = product_purchase_details.pricing[currDropdown.value];
        });
        //bind the button to the purchase process
        btn_buynow.addEventListener("click", function(){
          if (Agora._credentials.active
          && typeof product_purchase_details.pricing[currDropdown.value] != "undefined"
          && typeof Agora.balances[currDropdown.value] != "undefined"
          && Agora.balances[currDropdown.value] >= product_purchase_details.pricing[currDropdown.value]){
            //everything looks good to commence purchase
            blockingModal(true, "Please sign the purchase request to start minting");
            servercomms.sign_po(
              Agora._credentials.wallet, Agora._credentials.auth_token,
              product_purchase_details.name,
              productId,
              currDropdown.value,
              product_purchase_details.pricing[currDropdown.value]
            ).then(function(res){
              Agora.refresh_balances();
              if (res.status == "ok"){
                blockingModal(true, "Your collectable has been minted! Taking you to your Agora...");
                setTimeout(function(){
                  window.location = "/mine";
                }, 2000);
              }else{
                blockingModal(false);
                showToast("There was a problem minting your NFT. Your account has not been debited, please reload the page and try again.", "bad");
              }
            }).catch(function(err){
              blockingModal(false);
              showToast("NFT failed to mint.", "bad");
            });
          }else if (!Agora._credentials.active){
            showToast("Please login to purchase", "bad");
            document.getElementById("connect_btn").click();
          }else{
            showToast("Insufficient balance. Please deposit.", "bad");
            document.getElementById("mycredit").click();
          }
        });
      }else{
        console.error("Failed to retrieve pricing info");
      }
    }).catch(function(err){
      console.log(err);
    });
  });
}else{
  //This URL is incorrectly formatted
  console.error("Not proper URL");
}
